### PR TITLE
Redesign main UI

### LINKED ISSUE #143: Create custom Scaffold
Currently, most of the top-level screens show similar widget architecture: There's a `FancyAppBar` and a `SliverPadding` with the same values on every screen inside a `CustomScrollView`.

We can probably extract that pattern into a custom version of a scaffold. This would also allow it to move faster when doing future changes because the design could be changed at one single place.

### PR DESCRIPTION
**Pull request by MarcelGarus**
Fixes: #139, fixes: #143

Using lots of colors distracted from the actual content and made the design look old. So here's a cleaner, decluttered design. :tada:

The bottom navigation bar has no menu anymore. Instead, there are several items directly inside the navigation bar, which makes navigating much easier. This is in accordance to the semi-recent design trend of removing hamburger menus in favor of more direct navigation (as seen on YouTube, Instagram, Google Maps etc.).
The account and settings information went to the (top) app bar instead, where an account avatar is displayed on the top right, just like on the web client. The app bar gets a shadow as soon as the user scrolls.

Most of the content is now organized into cards to provide a consistent layout throughout the app.
To document the chosen padding (and future design decisions), there's now a [design wiki page](https://github.com/schul-cloud/schulcloud-flutter/wiki/Design).

Also, there are now several extension methods for making coding easier.

Here are some screenshots of the new design:

dashboard | dashboard further down, elevated app bar | news
---|---|---
image | image | image

courses | assignments | files
---|---|---
image | image | image

account dialog | dashboard in dark mode | news in dark mode
---|---|---
image | image | image

**Known issues**

* The course screen has overflowing courses. This is an issue in general with the current widget and will be fixed later. See #140.
* The app bar doesn't animate between screens. This is something to tackle in a future issue.
* The file browser doesn't use the new app bar yet, because it would lose the fantastic smooth transition. As soon as the new app bar supports transitions, we could migrate to the new app bar.
* The settings screen doesn't use the new app bar yet. See #105.
* The account dialog could get some better design. See #141.
* The active tab bar item is illegible in dark mode. The theme will probably be changed to fix that.
* The assignments screen isn't redesigned yet. @JonasWanke is working on that.
* The padding on the files screen is off. That's probably an issue with the `MediaQuery`'s padding and will be fixed as soon as we use a custom scaffold. See #143.
* File course items on the file screen overflow, but that's an issue for another time. See #142.

### CODE REVIEW

File: lib/app/app.dart [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
-export 'theming/thr.dart' show thrAppConfig;
+export 'theming_utils.dart';
 export 'utils.dart';
-export 'widgets/app_bar.dart' show AppBarActions;
 export 'widgets/buttons.dart';
 export 'widgets/empty_state.dart';
 export 'widgets/error_widgets.dart';
 export 'widgets/fade_in.dart';
+export 'widgets/fancy_app_bar.dart';
+export 'widgets/fancy_card.dart';
```
Comment (JonasWanke) [inline lib/app/app.dart]: We might want to remove the `fancy_` prefix

File: lib/app/widgets/navigation_bar.dart [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    widget.onNavigate(_activeScreen);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final s = context.s;
+    final theme = context.theme;
+
+    return BottomNavigationBar(
+      selectedItemColor: theme.primaryColor,
```
Comment (JonasWanke) [inline lib/app/widgets/navigation_bar.dart]: This evaluates to black using the dark theme which is not distinguishable from the bar itself

File: lib/course/widgets/course_detail_screen.dart [hunk 3]
```diff
@@ -14,73 +14,70 @@ class CourseDetailsScreen extends StatelessWidget {
...
+      body: CachedBuilder<List<Lesson>>(
+        controller: services.get<CourseBloc>().fetchLessonsOfCourse(course),
+        errorBannerBuilder: (_, error, st) => ErrorBanner(error, st),
+        errorScreenBuilder: (_, error, st) => ErrorScreen(error, st),
+        builder: (context, lessons) {
+          if (lessons.isEmpty) {
+            return EmptyStateScreen(
+              text: context.s.course_detailsScreen_empty,
+            );
+          }
```
Comment (JonasWanke) [inline lib/course/widgets/course_detail_screen.dart:39]: If the course doesn't contain any lessons, we don't even display the course's name and description? I don't think that's what we want

File: lib/course/widgets/course_detail_screen.dart [hunk 4]
```diff
@@ -14,73 +14,70 @@ class CourseDetailsScreen extends StatelessWidget {
...
   void _showLessonScreen({BuildContext context, Lesson lesson, Course course}) {
-    Navigator.of(context).push(MaterialPageRoute(
+    context.navigator.push(MaterialPageRoute(
       builder: (context) => LessonScreen(course: course, lesson: lesson),
     ));
   }
 
   @override
   Widget build(BuildContext context) {
     return Scaffold(
```
Comment (JonasWanke) [inline lib/course/widgets/course_detail_screen.dart:29]: Why not use a `FancyScaffold`?

File: lib/course/widgets/courses_screen.dart [hunk 5]
```diff
@@ -21,11 +21,19 @@ class CoursesScreen extends StatelessWidget {
...
             );
           }
-          return GridView.count(
-            childAspectRatio: 1.5,
-            crossAxisCount: 2,
-            children: <Widget>[
-              for (var course in courses) CourseCard(course: course),
+          return CustomScrollView(
+            slivers: <Widget>[
+              FancyAppBar(title: Text('Courses')),
```
Comment (JonasWanke) [inline lib/course/widgets/courses_screen.dart]: Translation missing

File: lib/course/widgets/courses_screen.dart [hunk 6]
```diff
@@ -21,11 +21,19 @@ class CoursesScreen extends StatelessWidget {
               text: context.s.course_coursesScreen_empty,
             );
           }
-          return GridView.count(
-            childAspectRatio: 1.5,
-            crossAxisCount: 2,
-            children: <Widget>[
-              for (var course in courses) CourseCard(course: course),
+          return CustomScrollView(
```
Comment (JonasWanke) [inline lib/course/widgets/courses_screen.dart:23]: Why not use a `FancyScaffold`?

File: lib/course/widgets/lesson_screen.dart [hunk 7]
```diff
@@ -83,43 +82,32 @@ class _LessonScreenState extends State<LessonScreen> {
   @override
   Widget build(BuildContext context) {
     return Scaffold(
```
Comment (JonasWanke) [inline lib/course/widgets/lesson_screen.dart:84]: Why not use a `FancyScaffold`?

File: lib/l10n/l10n.dart [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import 'package:flutter/widgets.dart';
+import 'package:schulcloud/generated/l10n.dart';
+
+export 'package:schulcloud/generated/l10n.dart';
+
+extension Localization on BuildContext {
+  S get s => S.of(this);
```
Comment (JonasWanke) [inline lib/l10n/l10n.dart]: What do you think about merging this definition into the `FancyContext` extension? Extension methods in not-yet-imported files are not being suggested by autocompletion—moving this into the `app` package could save us from manually specifying an import as `app` is already included in most files

File: lib/news/widgets/article_preview.dart [hunk 9]
```diff
@@ -32,20 +31,20 @@ class ArticlePreview extends StatelessWidget {
...
 
     return Provider<ArticleTheme>(
       create: (_) => ArticleTheme(darkColor: Colors.purple, padding: 16),
       child: Material(
-        elevation: 2,
-        borderRadius: BorderRadius.circular(8),
+        elevation: 1,
+        borderRadius: BorderRadius.circular(16),
         color: theme.cardColor,
         child: InkWell(
```
Comment (JonasWanke) [inline lib/news/widgets/article_preview.dart]: Can we use a `FancyCard` here and move it outside of `Provider`?

File: lib/news/widgets/news_screen.dart [hunk 10]
```diff
@@ -18,14 +18,22 @@ class NewsScreen extends StatelessWidget {
...
 
-          return ListView.builder(
-            itemCount: articles.length,
-            itemBuilder: (context, index) {
-              return Padding(
-                padding: EdgeInsets.fromLTRB(16, 16, 16, 0),
-                child: ArticlePreview(article: articles[index]),
-              );
-            },
+          return CustomScrollView(
```
Comment (JonasWanke) [inline lib/news/widgets/news_screen.dart:21]: Can we use a `FancyScaffold` here?

File: lib/app/widgets/fancy_card.dart [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+  Widget _buildContent(BuildContext context) {
+    return Column(
+      crossAxisAlignment: CrossAxisAlignment.start,
+      children: <Widget>[
+        SizedBox(height: 16),
+        if (title != null)
+          Padding(
+            padding: EdgeInsets.fromLTRB(16, 0, 16, 16),
+            child: Text(title.toUpperCase(), style: context.textTheme.overline),
```
Comment (JonasWanke) [inline lib/app/widgets/fancy_card.dart]: I think the title looks better with `mediumEmphasisColor`

File: lib/assignment/widgets/dashboard_card.dart [hunk 12]
```diff
@@ -2,18 +2,17 @@ import 'package:collection/collection.dart';
...
+import 'package:schulcloud/l10n/l10n.dart';
 
 import '../assignment.dart';
 import '../bloc.dart';
 
 class AssignmentDashboardCard extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
-    return DashboardCard(
+    return FancyCard(
```
Comment (JonasWanke) [inline lib/assignment/widgets/dashboard_card.dart:14]: We should probably use `omitHorizontalPadding`

**Review by JonasWanke — CHANGES_REQUESTED** [review]

**Comment by konstantin-edtech** [comment]
i like the style and glossyness in general. only thing i feel is that the tasks don't seem to be very clickable, because there is no link indicating colour, symbol, card or whatever. same for news on the dashboard. news on dashboard should look more similar to their look on the news.

**Comment by janrenz** [comment]
I always feel that cards are a waste of space on small screens. This place could be used for content and content parts can still be separated. See the new mobile design form Spiegel
https://apps.apple.com/de/app/der-spiegel/id424881832

**Comment by konstantin-edtech** [comment]
Let's agree on making the clickabilty of assignments more visible (Sarah dindn't get it too).

Remark from Sarah: Course cards could also be optimized to the new layout. For me as PO Prio 2, if low hanging then do it, if not iterate later

@dominikbruechner could the component libray be useful here? 
@marcelgarus @JonasWanke please contact Sarah in RocketChat to meet her for feedback. And do you know our story book? https://storybook.test.schul-cloud.org/?path=/story/base-base-ui--basebutton

**Comment by MarcelGarus** [comment]
@janrenz I do agree that cards are a waste of screen estate for large lists of homogenous content, like news content in the Spiegel app — images that go all the way from one side of the screen to the other look way more immersive than if there was padding around them.
For groups of heterogenous content, however, I believe that cards provide a great way of grouping related information together. For example, unlike news articles where basically the only types of content are text and images, on the dashboard of the Schul-Cloud app, there are schedules with progress bars, assignments and news and much more in the future and everything has a different layout. I believe that's a great fit for cards.
This distinction between normal lists for homogenous content and cards for heterogeneous content can be seen in lots of other apps to — for example, Twitter and Instagram, which provide feeds consisting mostly of text or images, respectively, don't use cards. Google search results, LinkedIn, Facebook instead use cards to group units of content together, because their feed can contain anything from polls to text to images to website previews, etc.

So, while I'd go along with using something other than cards on the news screen or the files screen, I'd strongly argue for using cards on the dashboard. If you're still not convinced, here's a beautiful render of how active events could get highlighted in the future (we still need to fix the padding):

image

@konstantin-sc About the assignments screen, we know it's very bare-bones for now — @JonasWanke is working on it, as pointed in the _known issues_ section in the PR description.

We didn't know the storybook yet, looks interesting. When we implement some more UI to talk about, we suggest having a design meeting with the UX team sometime in the following weeks.

**Review by JonasWanke — APPROVED** [review]

**Comment by janrenz** [comment]
Also reducing the shadow could make this look lighter, but I keep this ti the UX team.. however great work!

**Comment by MarcelGarus** [comment]
Yup, it's definitely better than what's currently in the master (design-wise and code-wise), so I'm gonna go ahead and merge this. Future design changes are welcome of course. 😊

**Comment by konstantin-edtech** [comment]
👍 UX Team had a short review. Further discussions in #mobile or best face to face.

@janrenz the truth for the spiegel app lies in the middle: on dashboard, they group connected items on a card (e.g. articles about corona virus or best of spiegel plus). single articles are on a single card.

**Pull request merged**